src/Microsoft.IdentityModel.Protocols.SignedHttpRequest/SignedHttpRequestHandler.cs

```diff
@@ -0,0 +1,1417 @@
+{
+    /// <summary>
+    /// A handler designed for creating and validating signed http requests. 
+    /// </summary>
+    /// <remarks>The handler implementation is based on 'A Method for Signing HTTP Requests for OAuth' specification.</remarks>
+    public class SignedHttpRequestHandler
+    {
+        // (https://tools.ietf.org/html/draft-ietf-oauth-signed-http-request-03#section-3.2)
+        // "Encodes the name and value of the header as "name: value" and appends it to the string buffer separated by a newline "\n" character."
+        private readonly string _newlineSeparator = "\n";
+
+        private readonly JsonWebTokenHandler _jwtTokenHandler = new JsonWebTokenHandler();
+        private readonly Uri _baseUriHelper = new Uri("http://localhost", UriKind.Absolute);
+        private readonly HttpClient _defaultHttpClient = new HttpClient();
+
+        #region SignedHttpRequest creation
+        /// <summary>
+        /// Creates a signed http request using the <paramref name="signedHttpRequestDescriptor"/>.
+        /// /// </summary>
+        /// <param name="signedHttpRequestDescriptor">A structure that wraps parameters needed for SignedHttpRequest creation.</param>
+        /// <param name="cancellationToken">Propagates notification that operations should be canceled.</param>
+        /// <returns>A signed http request as a JWS in Compact Serialization Format.</returns>
+        public async Task<string> CreateSignedHttpRequestAsync(SignedHttpRequestDescriptor signedHttpRequestDescriptor, CancellationToken cancellationToken)
+        {
+            if (signedHttpRequestDescriptor == null)
+                throw LogHelper.LogArgumentNullException(nameof(signedHttpRequestDescriptor));
+
+            var header = CreateHttpRequestHeader(signedHttpRequestDescriptor);
+            var payload = CreateHttpRequestPayload(signedHttpRequestDescriptor);
+            return await SignHttpRequestAsync(header, payload, signedHttpRequestDescriptor, cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Creates a JSON representation of a HttpRequest header.
+        /// </summary>
+        /// <param name="signedHttpRequestDescriptor">A structure that wraps parameters needed for SignedHttpRequest creation.</param>
+        /// <returns>A JSON representation of an HttpRequest header.</returns>
+        protected virtual string CreateHttpRequestHeader(SignedHttpRequestDescriptor signedHttpRequestDescriptor)
+        {
+            var header = new JObject
+            {
+                { JwtHeaderParameterNames.Alg, signedHttpRequestDescriptor.SigningCredentials.Algorithm },
+                { JwtHeaderParameterNames.Typ, SignedHttpRequestConstants.TokenType }
+            };
+
+            if (signedHttpRequestDescriptor.SigningCredentials.Key is X509SecurityKey x509SecurityKey)
+                header[JwtHeaderParameterNames.X5t] = x509SecurityKey.X5t;
```

Comment:
Curious:
Under what conditions should the x5t claim in the JWT header be used?

Reply:
This code is not used anymore and JsonWebToken handler is utilized instead: https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1260/commits/a1656abf4f653567a5d6e97a49aeed2559d331a5